tree/forest/v7/src/RPageStorageRoot.cxx

```diff
@@ -0,0 +1,355 @@
+            fMapper.fColumnIndex[iColumn].fPointeeClusterOffset.push_back(pointeeClusterOffset);
+            pageInCluster++;
+         }
+      }
+      delete clusterFooter;
+   }
+
+   for (unsigned iColumn = 0; iColumn < nColumns; ++iColumn) {
+      fMapper.fColumnIndex[iColumn].fNElements = forestFooter->fNElementsPerColumn[iColumn];
+   }
+   fMapper.fNEntries = forestFooter->fNEntries;
+
+   delete forestFooter;
+   delete forestHeader;
+
+   // TODO(jblomer): replace RMapper by a forest descriptor
+   RForestDescriptorBuilder descBuilder;
+   descBuilder.SetForest(fForestName, RForestVersion());
+   fDescriptor = descBuilder.GetDescriptor();
+}
+
+
+std::unique_ptr<ROOT::Experimental::RForestModel> ROOT::Experimental::Detail::RPageSourceRoot::GenerateModel()
+{
+   auto model = std::make_unique<RForestModel>();
+   for (auto& f : fMapper.fRootFields) {
+      auto field = Detail::RFieldBase::Create(f.fFieldName, f.fTypeName);
+      model->AddField(std::unique_ptr<Detail::RFieldBase>(field));
+   }
+   return model;
+}
+
+void ROOT::Experimental::Detail::RPageSourceRoot::PopulatePage(
+   ColumnHandle_t columnHandle, ForestSize_t index, RPage* page)
+{
+   auto columnId = columnHandle.fId;
+   auto nElems = fMapper.fColumnIndex[columnId].fNElements;
+   R__ASSERT(index < nElems);
+
+   ForestSize_t firstInPage = 0;
+   ForestSize_t firstOutsidePage = nElems;
+   ForestSize_t pageIdx = 0;
+
+   std::size_t iLower = 0;
+   std::size_t iUpper = fMapper.fColumnIndex[columnId].fRangeStarts.size() - 1;
+   R__ASSERT(iLower <= iUpper);
+   unsigned iLast = iUpper;
+   while (iLower <= iUpper) {
+      std::size_t iPivot = (iLower + iUpper) / 2;
+      ForestSize_t pivot = fMapper.fColumnIndex[columnId].fRangeStarts[iPivot];
+      if (pivot > index) {
+         iUpper = iPivot - 1;
+      } else {
+         auto next = nElems;
+         if (iPivot < iLast) next = fMapper.fColumnIndex[columnId].fRangeStarts[iPivot + 1];
+         if ((pivot == index) || (next > index)) {
+            firstOutsidePage = next;
+            firstInPage = pivot;
+            pageIdx = iPivot;
+            break;
+         } else {
+            iLower = iPivot + 1;
+         }
+      }
+   }
+
+   auto elemsInPage = firstOutsidePage - firstInPage;
+   void* buf = page->Reserve(elemsInPage);
+   R__ASSERT(buf != nullptr);
+
+   auto clusterId = fMapper.fColumnIndex[columnId].fClusterId[pageIdx];
+   auto pageInCluster = fMapper.fColumnIndex[columnId].fPageInCluster[pageIdx];
+   auto selfOffset = fMapper.fColumnIndex[columnId].fSelfClusterOffset[pageIdx];
+   auto pointeeOffset = fMapper.fColumnIndex[columnId].fPointeeClusterOffset[pageIdx];
+   page->SetWindow(firstInPage, RPage::RClusterInfo(clusterId, selfOffset, pointeeOffset));
+
+   //printf("Populating page %lu/%lu [%lu] for column %d starting at %lu\n", clusterId, pageInCluster, pageIdx, columnId, firstInPage);
+
+   std::string keyName = std::string(RMapper::kKeyPagePayload) +
```

Comment:
Isn't the same string creation happening in other place (i.e. opportunity for factoring)?